File: src/main/java/com/github/tanokun/tanorpg/game/entity/EntityManager.java
package com.github.tanokun.tanorpg.game.entity;

import com.github.tanokun.tanorpg.TanoRPG;
import com.github.tanokun.tanorpg.game.entity.base.ObjectEntity;
import com.github.tanokun.tanorpg.player.EquipmentMap;
import com.github.tanokun.tanorpg.player.status.StatusMap;
import com.github.tanokun.tanorpg.player.status.StatusType;
import com.github.tanokun.tanorpg.util.command.CommandContext;
import com.github.tanokun.tanorpg.util.io.Config;
import com.github.tanokun.tanorpg.util.io.Folder;
import org.bukkit.entity.Player;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.HashMap;

public class EntityManager {
    private HashMap<String, ObjectEntity> entities = new HashMap<>();

    public EntityManager(Player player){
        loadEntity(player);
    }

    public void loadEntity(Player p) {
        try {
            for (Config config : new Folder("mobs", TanoRPG.getPlugin()).getFiles()) {
                String name = config.getConfig().getString("name", "unknown");

                EntityType entityType = EntityType.valueOf(config.getConfig().getString("entityType", "ZOMBIE"));

                int level = config.getConfig().getInt("level", 0);

                int exp = config.getConfig().getInt("exp", 0);

                StatusMap statusMap = new StatusMap(); config.getConfig().getConfigurationSection("statuses").getKeys(false).forEach(text -> {
                    statusMap.addStatus(StatusType.valueOf(text), config.getConfig().getInt("statuses." + text, 0));
                });

                EquipmentMap equipMap = new EquipmentMap(); config.getConfig().getConfigurationSection("armors").getKeys(false).forEach(text -> {
                    equipMap.setEquip(EquipmentMap.EquipmentType.valueOf(text),
                            TanoRPG.getPlugin().getItemManager().getItem(config.getConfig().getString("armors." + text)).init(1));
                });

                EntityDropItems dropItems = new EntityDropItems(); final CommandContext cc = new CommandContext(null, new String[]{""});
                config.getConfig().getList("drops", new ArrayList<String>()).forEach(text -> {
                    String text2 = text + "";
                    cc.init(null, text2.split(" "));
                    dropItems.addDrop(
                            TanoRPG.getPlugin().getItemManager().getItem(cc.getArg(0, "")).init(Integer.valueOf(cc.getArg(1, "null"))),
                            Double.valueOf(cc.getArg(2, "null")));
                });
                ObjectEntity objectEntity;
                Constructor<? extends ObjectEntity> constructor =
                        entityType.getClazz().getConstructor(Config.class, String.class, StatusMap.class, EquipmentMap.class, EntityDropItems.class, int.class, int.class);
                objectEntity = constructor.newInstance(config, name, statusMap, equipMap, dropItems, exp, level);
                entities.put(name, objectEntity);

            }
        }catch (Exception e){
            if (p != null){
                p.sendMessage(TanoRPG.PX + "§cエンティティ系のコンフィグでエラーが発生しました。");
            } else {
                TanoRPG.getPlugin().getLogger().warning("エンティティ系のコンフィグでエラーが発生しました。");
            }
        }

        if (p != null){
            p.sendMessage(TanoRPG.PX + "§aエンティティ系のコンフィグをロードしました。");
        } else {
            TanoRPG.getPlugin().getLogger().warning("エンティティ系のコンフィグをロードしました。");
        }
    }

    public HashMap<String, ObjectEntity> getEntities() {
        return entities;
    }

    public ObjectEntity getEntity(String name){
        return entities.get(name);
    }
}

File: src/main/java/com/github/tanokun/tanorpg/game/entity/EntityManager.kt
package com.github.tanokun.tanorpg.game.entity

import com.github.tanokun.tanorpg.DataManager
import com.github.tanokun.tanorpg.TanoRPG
import com.github.tanokun.tanorpg.TanoRPG.Companion.plugin
import com.github.tanokun.tanorpg.game.entity.boss.BossEntity
import com.github.tanokun.tanorpg.game.entity.boss.BossManager
import com.github.tanokun.tanorpg.game.item.drop.ItemTable
import com.github.tanokun.tanorpg.player.status.StatusMap
import com.github.tanokun.tanorpg.player.status.StatusType
import com.github.tanokun.tanorpg.util.io.Config
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.io.File

class EntityManager(player: Player?) {
    val bossManager = BossManager()
    val entities = HashMap<String, ObjectEntity>()
    private val ids = HashMap<String, String>()

    init {
        if (player != null) player.sendMessage(TanoRPG.PX + "§bLoading mob configs...")
        else Bukkit.getConsoleSender().sendMessage("[TanoRPG] §bLoading mob configs...")
        loadEntity(player, false, Config( plugin.mythicMobs, "Mobs").file)
    }

    private fun loadEntity(p: Player?, b: Boolean, tempFile: File) {
        var path = ""
        var filePath = ""
        val errors = HashSet<String>()
        if (!b) errors.add("§a    Mobs configs loaded without errors.")
        val files = tempFile.listFiles()
        for (tmpFile in files) {
            if (tmpFile.isDirectory) {
                loadEntity(p, true, tmpFile)
            } else try {
                val config = Config(plugin, tmpFile, tmpFile.name)
                filePath = "..." + File.separator + tempFile.name + File.separator + config.fileName + File.separator
                for (id in config.config.getKeys(false)) {
                    if (!config.config.isSet("$id.TanoRPG_level")) continue
                    path = "$id.Display"
                    val name = config.config.getString(path, "unknown")!!
                    path = "$id.TanoRPG_level"
                    val level = config.config.getInt(path, 0)
                    path = "$id.TanoRPG_exp"
                    val exp = config.config.getInt(path, 0)
                    val statusMap = StatusMap()
                    path = "$id.TanoRPG_statuses"
                    for (text in config.config.getConfigurationSection(path)!!.getKeys(false)) {
                        path = "$id.TanoRPG_statuses.$text"
                        statusMap.addStatus(StatusType.valueOf(text), config.config.getDouble(path, 0.0))
                        path = "$id.TanoRPG_statuses"
                    }
                    path = "$id.TanoRPG_drops"
                    var lootTable: ItemTable?
                    if (config.config.isSet(path)) {
                        lootTable = plugin.dropManager.getItemTable(config.config.getString(path)!!)
                        requireNotNull(lootTable) { "ルートテーブルが存在しません" }
                    } else {
                        lootTable = ItemTable(name, 1, 1, ArrayList())
                    }
                    val objectEntity = ObjectEntity(id, name, exp, level, statusMap, lootTable)
                    if (config.config.isSet("$id.TanoRPG_Boss.spawnLoc")) {
                        path = "$id.TanoRPG_Boss.spawnLoc"
                        val spawn = DataManager.getLocation(config, path)
                        path = "$id.TanoRPG_Boss.teleportLoc"
                        val teleport = DataManager.getLocation(config, path)
                        path = "$id.TanoRPG_Boss.clearLoc"
                        val clear = DataManager.getLocation(config, path)
                        path = "$id.TanoRPG_Boss.chest"
                        val itemTable = plugin.dropManager.getItemTable(config.config.getString(path)!!)
                        val bossEntity = BossEntity(objectEntity, clear, teleport, spawn, itemTable)
                        bossManager.register(id, bossEntity)
                    }
                    ids[name] = id
                    entities[id] = objectEntity
                }
            } catch (e: Exception) {
                errors.remove("§a    Mobs configs loaded without errors.")
                errors.add("§c    " + e.message + "§7" + "(Path: " + filePath + path + ")")
            }
        }
        showErrors(errors, p)
    }


    private fun showErrors(errors: HashSet<String>, p: Player?) {
        if (p != null) errors.stream().forEach { e: String -> p.sendMessage(e)
        } else errors.stream().forEach { e: String -> Bukkit.getConsoleSender().sendMessage(e) }
    }

    fun getID(name: String): String? {
        return ids[name]
    }

    fun getEntity(name: String): ObjectEntity? {
        return entities[name]
    }

    companion object {
        var AI = "TanoRPG_AI"
        var DISPLAY_NAME = "TanoRPG_displayName"
        @JvmField
        var ENTITY = "TanoRPG_entity"
    }
}